lib/TPP/Conversion/ConvertVectorToXsmm/ConvertTranspose.cpp

```diff
@@ -0,0 +1,299 @@
+    unaryInfo.ldi = stridesOnInput->front();
+    auto stridesOnOutput = mlir::utils::getStaticStrides(output->getOperand(1));
+
+    // Adjust ldo based on the VNNI factor.
+    unaryInfo.ldo =
+        stridesOnOutput->front() /
+        *vnni::utils::getVnniBlockingFactor(expandShapeOp.getSrcType());
+    auto functionOp = transposeOp->getParentOfType<func::FuncOp>();
+    OpBuilder::InsertionGuard guard(rewriter);
+    rewriter.setInsertionPoint(&*functionOp.getBody().op_begin());
+
+    // If `OpTy` is unary or binary we need to dispatch and extra
+    // integer for the kind of operation to invoke.
+    xsmm::UnaryKindAttr kind =
+        xsmm::UnaryKindAttr::get(rewriter.getContext(), xsmm::UnaryKind::VNNI2);
+
+    dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
+        loc, integer64, cast<TypedAttr>(kind)));
+    dispatchOperandTypes.push_back(integer64);
+
+    dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
+        loc, integer64, cast<TypedAttr>(dtype)));
+    dispatchOperandTypes.push_back(integer64);
+
+    DenseI64ArrayAttr dims = DenseI64ArrayAttr::get(
+        rewriter.getContext(), ArrayRef<int64_t>{unaryInfo.m, unaryInfo.n,
+                                                 unaryInfo.ldi, unaryInfo.ldo});
+    for (auto idx = 0; idx < dims.size(); idx++) {
+      dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
+          loc, integer64, rewriter.getIntegerAttr(integer64, dims[idx])));
+      dispatchOperandTypes.push_back(integer64);
+    }
+
+    // Dispatch the flags. Pass to the library the already ored-flag to
+    // avoid changing the interface every time we add a new flag. Flags
+    // are assumed to be verified before (i.e., op verifier).
+    auto flags = rewriter.getArrayAttr(xsmm::UnaryFlagsAttr::get(
+        rewriter.getContext(), xsmm::UnaryFlags::NONE));
+    int64_t oredFlag = xsmm::utils::getOredFlags(flags);
+    dispatchOperands.push_back(rewriter.create<arith::ConstantOp>(
+        loc, integer64, IntegerAttr::get(rewriter.getI64Type(), oredFlag)));
+    dispatchOperandTypes.push_back(integer64);
+
+    auto dispatched = xsmm::utils::buildDispatchCall(
+        rewriter, loc, dispatchOperands, dispatchOperandTypes, module,
+        SymbolRefAttr::get(transposeOp->getContext(), dispatchName));
+
+    rewriter.setInsertionPoint(transposeOp);
+    SmallVector<Value> operandRange;
+    operandRange.push_back(rewriter.create<arith::ConstantOp>(
+        loc, integer64, cast<TypedAttr>(dtype)));
+    operandRange.push_back(dispatched.getResult(0));
+    operandRange.push_back(input->getOperand(0));
+    operandRange.push_back(output->getOperand(1));
+    SmallVector<Value> inputs;
+    SmallVector<Value> preceedingOperands;
+    auto invokeCall = xsmm::utils::buildInvokeCall(
+        rewriter, output, module, inputs, preceedingOperands, -1, operandRange,
+        invokeName, dtype, true);
+
+    return std::make_pair(&*dispatched, &*invokeCall);
```

Comment:
This is a repeat of the code below. Please condense the two transposes into one function with different arguments for VNNI and non-VNNI.